### PR TITLE
Reformatted get_contact()

### PR DESCRIPTION
**Pull request by Hypolite**
This is a free-floating PR related to #3213 without any behavior change.

It's an improvement of `get_contact()` that solves a `@todo` and reformat the function for readability and code standards compliance.

### CODE REVIEW

File: include/Contact.php [hunk 1]
```diff
@@ -508,72 +508,81 @@ function contacts_not_grouped($uid,$start = 0,$count = 0) {
...
+	$contact_id = 0;
 
-	if (!$contact)
-		$contact = q("SELECT `id`, `avatar-date` FROM `contact` WHERE `alias` IN ('%s', '%s') AND `uid` = %d ORDER BY `id` LIMIT 1",
-				dbesc($url),
-				dbesc(normalise_link($url)),
-				intval($uid));
+	// Catch-all query, may return multiple rows
+	$contacts = q("SELECT `id`, `avatar-date` FROM `contact`
+				WHERE ('%s' IN (`url`, `addr`, `alias`) OR '%s' IN (`nurl`, `alias`))
```
Comment (annando) [inline include/Contact.php]: This is a smart solution - but a huge performance killer. This most likely results in full table scans - especially at the public contacts where are very much entries.

File: include/Contact.php [hunk 2]
```diff
@@ -508,72 +508,81 @@ function contacts_not_grouped($uid,$start = 0,$count = 0) {
...
 		if (!$update_photo OR $no_update) {
-			return($contactid);
+			return $contact_id;
 		}
-	} elseif ($uid != 0)
+	} elseif ($uid != 0) {
+		return 0;
+	}
+
+	$data = probe_url($url);
```
Comment (annando) [inline include/Contact.php]: This you could replace with the new function name (I guess it was Probe::uri)

File: include/Contact.php (hunk 1 above)
Comment (Hypolite) [inline include/Contact.php]: I ran a few tests, with a 5,000-rows `contact` table, an unsuccessful query is 0.05s, while a successful query is 0.01s. I expect this version to be definitely faster than the original when supplied a `nick@server.tld` URL as argument, and on par when supplied an existing `http://server.tld/nick` URL.

Comment (annando) [inline include/Contact.php]: Have you used "explain" on that? The system can't chose one single index for the query that would fit for every condition. Although it is not that smart, it is much faster to split this query.

My system - for example - has 130,000 entries in the contact table for public contacts.

I just did the "explain". The system chooses the "uid_name" index - which means a table scan for all entries for the given uid - and a sample query just lasted 3.08 seconds. Even a successful query nearly lasts a full second on my system - while split queries are done in hundreds of a second.

File: include/Contact.php [hunk 3]
```diff
@@ -1,7 +1,5 @@
 <?php
 
-require_once('include/Probe.php');
```
Comment (annando) [inline include/Contact.php]: Why had you removed it from here? According to @fabrixxm we should do all the includes at the top of the files - while in the past we did the opposite.

I don't know which is better - but we should do this consistently.

File: include/Contact.php (hunk 1 above)
Comment (Hypolite) [inline include/Contact.php]: Gotcha, will split the query. My performance problem was with the extra `probe_url()` for the email-like profile URLs.

File: include/Contact.php (hunk 3 above)
Comment (Hypolite) [inline include/Contact.php]: I'm more and more pessimistic about us being able to move to a fully autoloaded codebase, so I figured I'd start improving where I can. Unnecessary cascading inclusion eats up to .4s for each request. For a full page it's not a big deal but for pings and updates it's a killer.

File: include/Contact.php (hunk 1 above)
Comment (annando) [inline include/Contact.php]: Yeah, the probing is really time intense and should be avoided when possible. When you split the query you only have to split it in three queries: One for the "nurl", one for the "addr" and one for the "alias". The "url" needn't to be searched (and there is no index on it) - but the "alias" has to be queried for either the normalized and unnormalized value.

File: include/Contact.php [hunk 4]
```diff
@@ -534,19 +534,20 @@ function get_contact($url, $uid = 0, $no_update = false) {
...
 
-	// We first try the addr (nick@server.tld)
+	// We first try the nurl (http://server.tld/nick), most common case
 	$contacts = q("SELECT `id`, `avatar-date` FROM `contact`
-				WHERE `addr` = '%s'
-				AND `uid` = %d",
-		dbesc($url),
-		intval($uid));
+					WHERE `nurl` = '%s'
+					AND `uid` = %d",
```
Comment (annando) [inline include/Contact.php]: The original query had contained a limit. Why had you removed that?

Comment (Hypolite) [inline include/Contact.php]: The first query had a `LIMIT 2`, and the case where multiple contacts are found for a single `uid` is covered later in the code. Basically, there was no reason to keep it.

File: include/Contact.php [hunk 5]
```diff
@@ -508,72 +507,100 @@ function contacts_not_grouped($uid,$start = 0,$count = 0) {
...
 		}
-	} elseif ($uid != 0)
+	} elseif ($uid != 0) {
+		// Non-existing user-specific contact, exiting
+		return 0;
+	}
+
+	require_once('include/Probe.php');
+	$data = Probe::uri($url);
+	if (!$data['url']) {
```
Comment (annando) [inline include/Contact.php]: I just saw that you changed the behavior here. Please have a look at the current code. When the given url doesn't belong to Friendica, Diaspora or OStatus it tries to fetch some data from the gcontact table. Only if this fails as well, we are quitting.

Comment (Hypolite) [inline include/Contact.php]: Ha, gotcha.

File: include/Contact.php [hunk 6]
```diff
@@ -599,6 +594,12 @@ function get_contact($url, $uid = 0, $no_update = false) {
 		$data = $gcontacts[0];
 	}
 
+	// Unable to convert nick@server.tld into http://server.tld/nick
```
Comment (annando) [inline include/Contact.php]: I guess that we don't need these lines. When the probing fails, the system will return NETWORK_PHANTOM as a network.

File: include/Contact.php [hunk 7]
```diff
@@ -577,11 +577,6 @@ function get_contact($url, $uid = 0, $no_update = false) {
...
 	require_once('include/Probe.php');
 	$data = Probe::uri($url);
-	if (!$data['url']) {
-		return 0;
-	}
-
-	$url = $data["url"];
 
 	// Does this address belongs to a valid network?
 	if (!in_array($data["network"], array(NETWORK_DFRN, NETWORK_OSTATUS, NETWORK_DIASPORA))) {
```
Comment (annando) [inline include/Contact.php]: I just realized that the probing now returns data for pump.io based networks as well. So I guess we should include this network here as well.

File: include/Contact.php (hunk 6 above)
Comment (Hypolite) [inline include/Contact.php]: Even better, thanks for pointing that out.

Comment (annando) [inline include/Contact.php]: Yeah. I guess this behavior isn't documented very well.

File: include/Contact.php [hunk 8]
```diff
@@ -508,72 +507,96 @@ function contacts_not_grouped($uid,$start = 0,$count = 0) {
...
+ * nick@server.tld. We quit if we can't get any info back.
+ *
+ * Third, we create the contact record if it doesn't exist
+ *
+ * Fourth, we update the existing record with the new data (avatar, alias, nick)
+ * if there's any updates
+ *
  * @param string $url Contact URL
- * @param integer $uid The user id for the contact
+ * @param integer $uid The user id for the contact (0 = public contact)
```
Comment (Hypolite) [inline include/Contact.php]: You're welcome, it made sense to me when I dug into the like issue.

**Pull request merged**